pkg/controller/daemon/daemon_controller.go

```diff
@@ -1198,6 +1198,10 @@ func (dsc *DaemonSetsController) updateDaemonSetStatus(ds *apps.DaemonSet, nodeL
 		return fmt.Errorf("error storing status for daemon set %#v: %v", ds, err)
 	}
 
+	// Resync the DaemonSet after MinReadySeconds as a last line of defense to guard against clock-skew.
```

Comment:
it would help to enhance this comment with a more detailed explanation of this issue. I am still not following how the nodes (on which pod is running)clock skew from controller nodes clock causes this ?

Reply:
Seems like the reason would be this line https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemon_controller.go#L1174 ,  the reason from what i understand is 
1: kubelet marks the pod ready and changes LastTransitionTime
2: controller checks IsPodAvailable() and finds the pod unavailable since its diffing LastTransitionTime(set by kubelet) with time.Now which is controller time. Since controller is behind in clock, minreadySeconds is not satisfied and it marks unavailable even though minReadySeconds is satisfied

can one of you confirm if this is the right understanding of this issue @DaiHao @kargakis  ?
Also in this case when a pod is marked unavailable, why wont it be requeued ?

I think including all of this explanation will be helpful

Reply:
see here. https://github.com/kubernetes/kubernetes/pull/77208#discussion_r279599055

Reply:
> Seems like the reason would be this line https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/daemon/daemon_controller.go#L1174 , the reason from what i understand is
> 1: kubelet marks the pod ready and changes LastTransitionTime
> 2: controller checks IsPodAvailable() and finds the pod unavailable since its diffing LastTransitionTime(set by kubelet) with time.Now which is controller time. Since controller is behind in clock, minreadySeconds is not satisfied and it marks unavailable even though minReadySeconds is satisfied
> 
> can one of you confirm if this is the right understanding of this issue @DaiHao @kargakis ?
> Also in this case when a pod is marked unavailable, why wont it be requeued ?
> 
> I think including all of this explanation will be helpful

Your understanding is right. 
Pod's status do not change in the sync loop, controller mark it unavailable only in daemonset's status, but which is equal with its last status, so daemonset also not requeue.
see here. https://github.com/kubernetes/kubernetes/blob/b219272a76bb4ad752417e310b5b50a6e8dc55c0/pkg/controller/daemon/daemon_controller.go#L1107-L1117